Tests/AK/TestFixedArray.cpp

```diff
@@ -24,3 +26,56 @@ TEST_CASE(ints)
     EXPECT_EQ(ints[1], 1);
     EXPECT_EQ(ints[2], 2);
 }
+
+TEST_CASE(swap)
+{
+    FixedArray<int> first = FixedArray<int>::must_create_but_fixme_should_propagate_errors(4);
+    FixedArray<int> second = FixedArray<int>::must_create_but_fixme_should_propagate_errors(5);
+    first[3] = 1;
+    second[3] = 2;
+    first.swap(second);
+    EXPECT_EQ(first.size(), 5u);
+    EXPECT_EQ(second.size(), 4u);
+    EXPECT_EQ(first[3], 2);
+    EXPECT_EQ(second[3], 1);
+}
+
+TEST_CASE(move)
+{
+    FixedArray<int> moved_from_array = FixedArray<int>::must_create_but_fixme_should_propagate_errors(6);
+    FixedArray<int> moved_to_array(move(moved_from_array));
+    EXPECT_EQ(moved_to_array.size(), 6u);
+    EXPECT_EQ(moved_from_array.size(), 0u);
+}
+
+TEST_CASE(no_allocation)
+{
+    FixedArray<int> array = FixedArray<int>::must_create_but_fixme_should_propagate_errors(5);
+    EXPECT_NO_CRASH("Assigments", [&] {
+        NoAllocationGuard guard;
+        array[0] = 0;
+        array[1] = 1;
+        array[2] = 2;
+        array[4] = array[1];
+        array[3] = array[0] + array[2];
+        return Test::Crash::Failure::DidNotCrash;
+    });
+
+    EXPECT_NO_CRASH("Move", [&] {
+        FixedArray<int> moved_from_array = FixedArray<int>::must_create_but_fixme_should_propagate_errors(6);
+        NoAllocationGuard guard;
+        FixedArray<int> moved_to_array(move(moved_from_array));
+        // We need to ensure that this destructor runs before the FixedArray destructor.
+        guard.~NoAllocationGuard();
```

Comment:
Same here regardin manual dtor call

Reply:
Unfortunately, it's not possible because `moved_to_array` would be destructed in the inner scope, still requiring me to manually ensure the early destruction of the guard.

```c++
NoAllocationGuard guard;
// construction without allocation: we can't have this in the outer unprotected scope.
FixedArray<int> moved_to_array(move(moved_from_array));
// if the scope ended here, we'd still have to ensure that one of the two scope variables (guard) destructs before the other (moved_to_array)
guard.~NoAllocationGuard();
return Test::Crash::Failure::DidNotCrash;
// destruction happens here, and afaik stack variable destruction doesn't have a specified order
```

Reply:
Alright, let's hope noone uses it with a guard like this in regular code though :acid2:

Reply:
That's not accurate, as destructors are called in first constructed last destructed order. https://isocpp.org/wiki/faq/dtors#order-dtors-for-locals

So you are guaranteed by language semantics that the fixed array will be destroyed before the guard.

Reply:
It's not really obvious what is even being tested here, it seems like you want to make sure you don't try to allocate or free during move? Isn't validating one kind of proof of the other?